[로또] Step3,4 제출하겠습니다

SEUNGBOONG opened the pull request:
5/29 
[Collections.singletonList](https://www.korecmblog.com/blog/arrays-asList-alternative-Collections-singletonList)
[assertAll](https://sas-study.tistory.com/316) 
참고해서 코드리뷰 수정했습니다.
컬렉션 싱글톤리스트를 사용하는 이유를 굳이 말하자면 메모리 사용량이라고 하는데 
이것도 데이터구조에서 말하는 O(n) 과 관련된 것일까요 >?
혹시 로또 끝나고 사다리 가기전에 한번 제 코드를 분석해보는 시간을 가져보고, 이론적인 부분 및 사다리에서 써야할 것들을 한번  둘러 보는 시간을 가져도 괜찮을까요?

5/28 
TDD기반 개발을 했어야했는데, 
TDD를 지키지 않고 했음을 반성하며, Test 습관화 하도록 하겠습니다. 
혹시 InputView 테스트를 할때 목을 사용해서 해야하는지 의문이 들고있습니다. 
일단 언젠가는 써야하므로 따로 공부를 진행중에 있기는 합니다. 
감사합니다.





5/27 
버그 수정 및 step4 진행 완료했습니다. 객체를 잘나눠 추가할떄 어려움이 크지않아 좋았 던 것 같습니다. 
질문이 있는데, 혹시 수동로또 넘버를 일급컬렉션으로 따로 관리한후 로또티켓에 데이터를 보내서 합치게 되면 
로또티켓은 두 가지 책임을 지는 것 같아서 단일책임원칙에 어긋나는 것 같습니다. 혹시 이것을 컨트롤러에서 
합치게되면 단일책임원칙을 어긋나지 않게 되는걸까요? 

5/25
enum 사용을 처음해봤는데, step1부터 했으면 더욱 좋았을것을 아쉬움이 남았습니다.
step2에서 썼던 네이밍들을 다시 정리해보는 시간을 가졌고,
리드미 전반 적인 수정을 했습니다.(enum추가떄매)
다행히 객체를 처음부터 최대한 잘나눠가며 진행해서 인지 메서드 추가하고 구현하는데 어려움이 크진 않았습니다.
그래서 객체지향의 중요성에 대해 한 번더 꺠닫고 가는 시간을 가진것 같습니다.

다음 제출은 step4와 버그 수정함께진행하겠습니다

SEUNGBOONG committed:
-

SEUNGBOONG committed:
Feat:
- 보너스볼 입력 메서드 구현

SEUNGBOONG committed:
Refactor:
- 전체적인 네이밍 수정
- 메서드 추가됨에따라 새로운 lottoRate 삽입
- converToDto stream사용

SEUNGBOONG committed:
Refactor:
- 전체적인 코드 수정이 있었다 .
- Enum을 사용함에 따라 새로운 메서드로 구현을했다 .

SEUNGBOONG committed:
Fix:
- toString 사용하지 않아도 제대로 값이 나와서  버그수정

SEUNGBOONG committed:
Fix:
- toString 사용하지 않아도 제대로 값이 나와서  버그수정

SEUNGBOONG committed:
Refactor:
-전체적인 메서드를 클린 코드를 위해 살짝 손을 봤다.

SEUNGBOONG committed:
Refactor:
- else 를 사용하는 요구사항을 지키지 않아서 전체적인 수정
- enum 사용을 통한 수정도 있다.

SEUNGBOONG committed:
Refactor:
- 클린 코드 지키기

SEUNGBOONG committed:
Feat:
- 전체적인 enum 구현

SEUNGBOONG committed:
docs :
- enum 추가로 인한 전체적인 README.md 수정 진행

SEUNGBOONG committed:
Feat:
- 새로운 메서드 추가
- checkContain 으로 메서드 이름 변경
- checkBonusCount -> 새로운 메서드 생성

SEUNGBOONG committed:
Fix:
- 쓸모없어진 클래스 삭제

SEUNGBOONG committed:
Fix :
- 보너스볼 여러 개 나오는 이슈 해결

SEUNGBOONG committed:
Fix :
- 보너스볼 여러 개 나오는 이슈 해결

SEUNGBOONG committed:
Feat:
- 수동 구매 로또 수, 번호 입력 메서드 구현

SEUNGBOONG committed:
Feat:
- 수동 로또 넘버 나오게 구현

SEUNGBOONG committed:
-Feat :
일급컬렉션으로 수동로또넘버 구현

SEUNGBOONG committed:
Fix :
- 수동 카운트만큼 자동나오도록 수정

SEUNGBOONG committed:
Docs:
- 구현 완료

SEUNGBOONG committed:
Docs:
- 구현 완료

SEUNGBOONG committed:
Refator:
- 수동 로또 번호 메서드 셋팅

SEUNGBOONG committed:
Refator:
- 수동로또 메서드 구현

SEUNGBOONG committed:
Refator:
- 처음 접근했던 수동로또 일급컬렉션 삭제

SEUNGBOONG committed:
Fix:
- prviate 위치변경

SEUNGBOONG committed:
Fix:
- prviate 위치변경

SEUNGBOONG committed:
Fix:
- prviate 위치변경

sosow0212 commented on README.md:
```diff
@@ -0,0 +1,89 @@
+기능 요구사항
```
현재 타겟 브랜치가 `nest-step:main`으로 되어있습니다! 자기 깃허브 아이디로 바꿔주세요~

SEUNGBOONG replied:
완료했습니다.

sosow0212 commented on src/main/java/controller/LottoController.java:
```diff
@@ -0,0 +1,61 @@
+package controller;
+
+import domain.*;
```
import의 와일드카드(*)는 지양해주세요!
[링크](https://velog.io/@hooni_/%EC%99%80%EC%9D%BC%EB%93%9C-%EC%B9%B4%EB%93%9C-import)

SEUNGBOONG replied:
넵

sosow0212 commented on src/main/java/controller/LottoController.java:
```diff
@@ -0,0 +1,61 @@
+
+    public static void main(String[] args) {
+        new LottoController().run();
+    }
```
상위 클래스에서 `Controller.run()`하도록 변경해주세요!

SEUNGBOONG replied:
혹시 이부분을 어떻게 적용해야 하는지 알수 있을까요?

sosow0212 commented on src/main/java/controller/LottoController.java:
```diff
@@ -0,0 +1,61 @@
+        winningLotto(tickets, money);
+    }
+
+    private void winningLotto(List<LottoTicket> tickets, int money) {
```
메서드 명은 동사로 시작해주세요~

SEUNGBOONG replied:
넵

sosow0212 commented on src/main/java/controller/LottoController.java:
```diff
@@ -0,0 +1,61 @@
+        outputView.LottoRateOfResult(rateOfReturn);
+    }
+
+    private void winning(List<LottoTicket> tickets, WinningLotto winningLotto, int bonusNumber) {
```
여기도 동사로 시작해주세요!

sosow0212 commented on src/main/java/controller/LottoController.java:
```diff
@@ -0,0 +1,61 @@
+        return tickets.stream()
+                .map(ticket -> new LottoTicketDto(ticket.getNumbers()))
+                .collect(Collectors.toList());
+    }
```
👍

저는 Dto에 정적 팩토리 메서드로 도메인을 넣고, Dto 내부에서 객체를 dto로 바꾸는 함수를 만들어서 사용하고 있습니다!
dto가 객체를 의존하는 문제가 있지만 더 간편하더라고요. 이런 방법도 있다 정도로 생각하고 편한 방식대로 쓰시면 좋을 것 같아요~

sosow0212 commented on src/main/java/domain/LottoRate.java:
```diff
@@ -0,0 +1,34 @@
+
+public class LottoRate {
+
+    private final List<LottoTicket> tickets;
+    private final WinningLotto winningLotto;
```
일급 컬렉션에 다른 필드가 있으면 안됩니다!

이렇게 사용하고 싶다면 `private final List<LottoTicket> tickets;` 이 부분을 `Tickets` 로 포장을 한 후, LottoRate에서 `Tickets, WinningLotto`를 사용하는게 좋을 것 같아요!

sosow0212 commented on src/main/java/domain/LottoRate.java:
```diff
@@ -0,0 +1,34 @@
+        this.winningLotto = winningLotto;
+    }
+
+    public List<Integer> matchesMoney() {
+        return List.of(
+                Ranking.FIRST.getPrice(),
+                Ranking.SECOND.getPrice(),
+                Ranking.THIRD.getPrice(),
+                Ranking.FOURTH.getPrice(),
+                Ranking.FIFTH.getPrice()
+        );
```
match를 해주는 것 같은데, 등수 별 금액을 반환하고 있습니다.
이 부분 자체도 Enum에서 해결해줄 수 있지 않을까요?

sosow0212 commented on src/main/java/domain/LottoRate.java:
```diff
@@ -0,0 +1,34 @@
+        );
+    }
+
+    public double calculateRateOfReturn(int payMoney, int lottoBonusNumber) {
+        int totalWinnings = tickets.stream()
+                .mapToInt(ticket -> {
+                    int matchedNumbers = winningLotto.getCount(ticket.getNumbers());
+                    int matchedBonusNumbers = winningLotto.getBonusCount(ticket.getNumbers(), lottoBonusNumber);
+                    return Ranking.valueOfCount(matchedNumbers, matchedBonusNumbers).getPrice();
+                }).sum();
+        return (double) totalWinnings / payMoney;
```
👍

sosow0212 commented on src/main/java/domain/LottoTicket.java:
```diff
@@ -0,0 +1,16 @@
+
+public class LottoTicket {
+
+    private final List<Integer> numbers;
```
List<Integer>에서 Integer도 요구사항에 맞게 원시값 포장을 해보면 어떨까요?

sosow0212 commented on src/main/java/domain/LottoTickets.java:
```diff
@@ -0,0 +1,35 @@
+
+public class LottoTickets {
+
+    private static final int INITIAL_NUMBER = 0;
+    private final NumberGenerator numberGenerator;
```
```suggestion
    private static final int INITIAL_NUMBER = 0;
    
    private final NumberGenerator numberGenerator;
```

일급 컬렉션에 다른 필드가 있으면 안됩니다!
NumberGenerator도 매개변수로 받아서 사용하는 건 어떨까요?

sosow0212 commented on src/main/java/domain/LottoTickets.java:
```diff
@@ -0,0 +1,35 @@
+        }
+    }
+
+    public List<LottoTicket> generateLottoTickets(int ticketCount) {
+        generateLottoTickets(ticketCount, tickets);
+        return tickets;
+    }
+
+    private void generateLottoTickets(int ticketCount, List<LottoTicket> tickets) {
+        for (int i = INITIAL_NUMBER; i < ticketCount; i++) {
+            List<Integer> numbers = numberGenerator.generateRandomNumber();
+            LottoTicket ticket = new LottoTicket(numbers);
+            tickets.add(ticket);
+        }
+    }
```
👍

sosow0212 commented on src/main/java/domain/LottoTickets.java:
```diff
@@ -0,0 +1,35 @@
+        this.tickets = new ArrayList<>();
+    }
+
+    public void addPassiveTickets(List<List<Integer>> passiveNumbers) {
+        for (List<Integer> numbers : passiveNumbers) {
+            tickets.add(new LottoTicket(numbers));
+        }
+    }
```
이 부분이 수동 구매 부분이네요!

객체로 빼는 것과 현재 방식의 차이는 무엇이라고 생각하시나요?

sosow0212 commented on src/main/java/domain/PassiveLottoTicket.java:
```diff
@@ -0,0 +1,11 @@
+//package domain;
```
안쓰는 클래스는 모두 제거해주세요~

sosow0212 commented on src/main/java/domain/Ranking.java:
```diff
@@ -0,0 +1,46 @@
+
+public enum Ranking {
+
+    FIRST(6, 0, 2000000000, "6개 일치 1등입니다."),
+    SECOND(5, 1, 30000000, "5개 일치, 보너스 볼 일치 2등입니다."),
+    THIRD(5, 0, 1500000, "5개 일치 3등입니다."),
+    FOURTH(4, 0, 50000, "4개 일치 4등입니다."),
+    FIFTH(3, 0, 5000, "3개 일기 5등입니다."),
+    ERROR(0, 0, 0, "");
```
1. ㅋㅋ FIFTH에 메시지가 오타가 있습니다.
2. 일치 메시지를 Enum에서 관리하면 결국 객체가 뷰를 의존하는 거 아닐까요??

sosow0212 commented on src/main/java/domain/Ranking.java:
```diff
@@ -0,0 +1,46 @@
+        return message;
+    }
+
+    public static Ranking valueOfCount(int countOfRanking, int bonusCountOfRanking) {
+        for (Ranking ranking : Ranking.values()) {
+            Ranking rankings = getRanking(countOfRanking, bonusCountOfRanking, ranking);
+            if (rankings != null) return rankings;
+        }
+        return ERROR;
+    }
+
+    private static Ranking getRanking(int countOfRanking, int bonusCountOfRanking, Ranking ranking) {
+        if (countOfRanking == ranking.countOfRanking && bonusCountOfRanking == ranking.bonusCountOfRanking) {
+            return ranking;
+        }
+        return null;
+    }
```
values().stream()
.filter(조건)
.findAny()
.orElseThrow()

이런식으로 스트림으로 한 번에 할 수 있어 보입니다 ㅎㅎ

sosow0212 commented on src/main/java/domain/TicketCount.java:
```diff
@@ -0,0 +1,16 @@
+
+    private final int money;
+
+    public TicketCount(int money, int count) {
+        this.money = money - (count * TICKET_PRICE);
+    }
+
+    public int getCount() {
+        return money / TICKET_PRICE;
+    }
+}
```
생성자는 수동구매 때문에 나눈 것인가요?

money는 그냥 입력 받고, 수동구매를 한다면 `money -= 금액`을 해주고
자동 구매도 `money -= 금액`을 해주면 수동구매와 자동구매의 순서가 바뀌더라도 구매가 가능할 것 같습니다.

그리고 0원이라면 구매하지도 못할 것 같은데 혹은 수동구매시 금액이 부족해도 예외가 발생할 것 같은데 처리해주면 좋을 것 같아요

SEUNGBOONG replied:
수정완료했습니다.

sosow0212 commented on src/main/java/domain/WinningLotto.java:
```diff
@@ -0,0 +1,39 @@
+    }
+
+    private int checkContain(int number, int count) {
+        if (winNumbers.contains(number)) count++;
```
```suggestion
        if (winNumbers.contains(number)) {
               count++;
       }
```

sosow0212 commented on src/main/java/domain/WinningLotto.java:
```diff
@@ -0,0 +1,39 @@
+        return count;
+    }
+
+    private int checkBonusCount(List<Integer> ticketNumbers, int bonusNumber, int bonusCount) {
+        if (ticketNumbers.contains(bonusNumber)) {
+            bonusCount++;
+        }
+        return bonusCount;
+    }
```
```suggestion
    private int checkBonusCount(List<Integer> ticketNumbers, int bonusNumber) {
        int bonusCount = 0;
        if (ticketNumbers.contains(bonusNumber)) {
            bonusCount++;
        }
        return bonusCount;
    }
```

sosow0212 commented on src/main/java/view/InputView.java:
```diff
@@ -0,0 +1,50 @@
+
+public class InputView {
+
+    private static final String COMMA = ",";
+    private static final int INITIAL_NUMBER = 0;
+    private final Scanner scanner = new Scanner(System.in);
+    private int count = 0;
```
```suggestion
    private static final String COMMA = ",";
    private static final int INITIAL_NUMBER = 0;
    
    private final Scanner scanner = new Scanner(System.in);
    
    private int count = 0;
```

sosow0212 commented on src/test/java/domain/LottoTest.java:
```diff
@@ -0,0 +1,49 @@
+import static org.junit.jupiter.api.Assertions.assertEquals;
+import static org.junit.jupiter.api.Assertions.assertTrue;
+
+class LottoTest {
```
다른 도메인도 모두 테스트 진행해주세요!

sosow0212 reviewed (changes requested):
오 구조적으로 많은 고민을 하신게 보입니다!
몇 가지 수정만 해주시면 될 것 같아요.

그리고 현재 테스트가 많이 비어있는데, 모두 채워주세요~
타겟 브랜치도 main으로 되어 있는데 자기 깃 이름으로 바꿔주시면 될 것 같습니다!!

궁금한 점 있으면 말해주세요~ 수고하셨습니다 👍

SEUNGBOONG committed:
Fix:
- 와일드카드 컨벤션 수정

SEUNGBOONG committed:
Fix:
- 원시값 포장으로 TicketCount를 관리하기 맞는지 의문이드는중

SEUNGBOONG committed:
Fix:
- 로또 테스트 통과

SEUNGBOONG committed:
Fix:
- 로또 테스트

SEUNGBOONG committed:
Feat:
- 예외처리 추가ㅓ

SEUNGBOONG committed:
Feat:
- Mock라이브러리추가

SEUNGBOONG committed:
Feat:
- 당첨금 테스트
- 수익률 테스트 구현

SEUNGBOONG committed:
Feat:
- 당첨번호 조회테스트
- 보너스볼 숫자세기 테스트

SEUNGBOONG committed:
Feat:

SEUNGBOONG committed:
Feat:
- 생성자 scanner 주입 후 Mock

sosow0212 commented on README.md:
```diff
@@ -25,27 +25,53 @@
 
 Model
-CountLotto
+LottoRate
```
기능구현목록 아주 좋습니다!

sosow0212 commented on src/main/java/domain/LottoTickets.java:
```diff
@@ -5,20 +5,31 @@
 public class LottoTickets {
 
-    private static final int INITIAL_NUMBER = 0;
+    private static final Integer INITIAL_NUMBER = 0;
     private final NumberGenerator numberGenerator;
```
```suggestion
    private static final Integer INITIAL_NUMBER = 0;
    
    private final NumberGenerator numberGenerator;
```

sosow0212 commented on src/main/java/domain/LottoRate.java:
```diff
@@ -4,20 +4,31 @@
-    private static final int FIVE_MATCHES = 150000;
-    private static final int SIX_MATCHES = 20000000;
-    private int sum = 0;
+    private final List<LottoTicket> tickets;
+    private final WinningLotto winningLotto;
```
요 부분 아직 변경이 안 되어있습니다 !

sosow0212 commented on src/main/java/domain/LottoTickets.java:
```diff
@@ -5,20 +5,31 @@
 public class LottoTickets {
 
-    private static final int INITIAL_NUMBER = 0;
+    private static final Integer INITIAL_NUMBER = 0;
```
Integer인 이유가 있을까요?

SEUNGBOONG replied:
자동완성때 잘못 선언한것 같습니다.

sosow0212 commented on src/main/java/domain/PassiveLottoTicket.java:
```diff
@@ -0,0 +1,11 @@
+//package domain;
```
여기도 반영이 안 되어 있습니다!

SEUNGBOONG replied:
passiveLottoTIcekt은 삭제한 클래스였습니다. 
삭제한것은 커밋해야하는줄 몰랐습니다. 다음부턴 삭제된것도 하겠습니다.

sosow0212 commented on src/main/java/domain/PassiveLottoTickets.java:
```diff
@@ -0,0 +1,20 @@
+//package domain;
```
해당 클래스도 지워주세요~

sosow0212 commented on src/main/java/domain/RandomNumberGenerator.java:
```diff
@@ -7,23 +7,24 @@
-    private static void checkDuplicateNumber(List<Integer> number) {
+    private static void checkDuplicate(List<Integer> numbers) {
         int randomNumber = random.nextInt(MAX) + COUNT_NUMBER;
-        if (!number.contains(randomNumber)) {
-            number.add(randomNumber);
+        if (!numbers.contains(randomNumber)) {
+            numbers.add(randomNumber);
         }
```
중복 확인 + 숫자 삽입

총 두가지의 역할을 하는 것 같아요.
메서드의 역할을 분리해보는 건 어떨까요?

sosow0212 commented on src/main/java/domain/Ranking.java:
```diff
@@ -0,0 +1,46 @@
+    SECOND(5, 1, 30000000, "5개 일치, 보너스 볼 일치 2등입니다."),
+    THIRD(5, 0, 1500000, "5개 일치 3등입니다."),
+    FOURTH(4, 0, 50000, "4개 일치 4등입니다."),
+    FIFTH(3, 0, 5000, "3개 일기 5등입니다."),
```
여기도 아직 반영이 안되어있습니다!

sosow0212 commented on src/main/java/domain/TicketCount.java:
```diff
@@ -4,13 +4,47 @@ public class TicketCount {
+        return buyMoney.getAmount();
+    }
+
+    public static class BuyMoney {
```
내부 클래스를 만든 이유가 있을까요?

sosow0212 commented on src/main/java/domain/TicketCount.java:
```diff
@@ -4,13 +4,47 @@ public class TicketCount {
+    }
+
+    public static class BuyMoney {
+        private final int amount;
+        private static final int EXCEPTION_INITIAL_NUMBER = 0;
```
```suggestion
        private static final int EXCEPTION_INITIAL_NUMBER = 0;

        private final int amount;
```

sosow0212 commented on src/main/java/domain/TicketCount.java:
```diff
@@ -4,13 +4,47 @@ public class TicketCount {
+        private static final int EXCEPTION_INITIAL_NUMBER = 0;
+
+        public BuyMoney(int amount, int count) {
+            if (amount <= EXCEPTION_INITIAL_NUMBER) {
+                throw new RuntimeException("로또를 구매할 돈은 음수가 될 수 없습니다.");
+            }
+            this.amount = amount - (TICKET_PRICE * count);
```
돈이 부족해서 사지 못하는 경우는 어떻게 될까요?

sosow0212 commented on src/main/java/view/InputView.java:
```diff
@@ -1,26 +1,54 @@
 public class InputView {
 
-    private static final String DISUNITE_COMMA = ",";
+    private Scanner scanner;
+    private static final String COMMA = ",";
+    private static final int INITIAL_NUMBER = 0;
+    private int count = 0;
```
```suggestion
    private static final String COMMA = ",";
    private static final int INITIAL_NUMBER = 0;
    
    private final Scanner scanner;
    
    private int count = 0;
```

sosow0212 replied:
count 변수가 필요한 이유가 있을까요?

sosow0212 commented on src/main/java/view/OutputView.java:
```diff
@@ -1,41 +1,27 @@
-            System.out.printf("%d개 일치 (%d원)\n", count, matchesMoney.get(INITIAL_NUMBER_THREE));
-        } else {
-            System.out.printf("%d개 일치\n", count);
+    public void NumberOfWinning(Ranking ranking) {
```
뷰가 도메인(Ranking)을 의존하고 있습니다!

sosow0212 commented on src/test/java/domain/LottoRateTest.java:
```diff
@@ -0,0 +1,41 @@
+
+    @Test
+    void 로또상금_테스트() {
+
```
```suggestion
```

다른 부분도 확인해주세요~

sosow0212 commented on src/test/java/domain/LottoRateTest.java:
```diff
@@ -0,0 +1,41 @@
+    void 로또상금_테스트() {
+
+        // given
+        LottoRate lottoRate = new LottoRate(null, null);
```
첫 값에 null이 들어갈 때 다른 문제는 없을까요?
LottoRate 객체 자체에 문제가 있진 않을까 고민해보면 좋을 것 같습니다

sosow0212 commented on src/test/java/domain/LottoRateTest.java:
```diff
@@ -0,0 +1,41 @@
+        LottoRate lottoRate = new LottoRate(tickets, winningLotto);
+
+        // when
+        double rateOfReturn = lottoRate.calculateRateOfReturn(5000, 7);
```
5000, 7 값들은 어디서 나온 값들인가요?
tickets와 winningLotto의 List<> 요소를 넣어서 비교하게 넣을 순 없을까요?

SEUNGBOONG replied:
5000원은 로또 구입금액입니다. 
7은 보너스 볼 넘버입니다. 
given에서 네이밍을 알기  쉽게 하고 쓰라고 하시는 말씀이실까요>?

sosow0212 replied:
오프라인으로 설명드리겠습니다.1

sosow0212 commented on src/test/java/domain/TicketCountTest.java:
```diff
@@ -0,0 +1,38 @@
+            new TicketCount.BuyMoney(amount, count);
+        });
+    }
+}
```
끝 개행 추가해주세요~

sosow0212 commented on src/test/java/domain/TicketCountTest.java:
```diff
@@ -0,0 +1,38 @@
+
+        // when
+        TicketCount.BuyMoney buyMoney = new TicketCount.BuyMoney(amount, count);
+        int ActualValue = buyMoney.getAmount();
```
1. actualValue 소문자 시작으로 바꿔주세요!
2. expectedValue는 given으로 올려주세요

SEUNGBOONG replied:
넵 수정했습니다.

sosow0212 commented on src/test/java/domain/WinningLottoTest.java:
```diff
@@ -0,0 +1,44 @@
+        int bonusCount2 = winningLotto.getBonusCount(ticketNumber2, bonusNumber);
+
+        // then
+        Assertions.assertEquals(1, bonusCount2);
+        Assertions.assertEquals(0, bonusCount1);
```
다중 값 검증을 할 땐 softly 혹은 assertAll을 이용하는 건 어떨까요?!

SEUNGBOONG replied:
[assertAll](https://sas-study.tistory.com/316) 보고 수정했습니다.

sosow0212 reviewed (changes requested):
반영되지 않은 리뷰들이 있어서 한 번 쉬고 가겠습니다!
pr 메시지를 보니 객체지향적으로 작성해서 리팩토링 시 크게 바뀌지 않았다는게 아주 인상깊네요 👍

5/27일자 답변: "단일책임원칙은 하나의 객체가 하나의 책임을 가져야한다."로 이해해주시면 좋을 것 같습니다. 그 책임을 수행하기 위해 여러 역할(메서드)을 할 수 있고요!
그렇다면 객체지향의 사실과 오해 책에서 자주 나온 협력 클래스는 현재 코드에서 어떤 클래스일까요? 이 부분도 고민해보면 좋을 것 같습니다!!

5/28일자 답변 : TDD가 항상 정답이라고 생각하지는 않지만 할 줄은 알아야된다고 생각합니다! 저도 TDD로 할 때도 있고 아닐 때도 있는데 복잡한 로직이 들어가는 프로그램이라면 도메인 별로 실수하지 않고 만들어서 구조적으로도, 구현적으로도 모두 좋은 것 같습니다.
현재 미션들을 하면서 TDD를 익히기 좋다고 생각하는데 다음 미션부터 한 번 해보면 좋을 것 같습니다!

그리고 Controller는 지금 테스트 할 필요는 없지만, 만약 한다면 Mock을 사용해볼 수도 있을 것 같습니다. 어떻게 테스트 하냐에 따라 다르겠지만, 만약 행위 검증을 한다고 하면 InputView를 Mock으로 두거나 Interface로 InputView를 만들고 전략패턴을 통해 행위 검증을 수행할 수 있을 것 같습니다.

이는 현재 상황에 맞게 필요한 방법을 선택하면 될 것 같습니다!

SEUNGBOONG committed:
Feat:
- 가상 값 테스트 assertAll로 변경

SEUNGBOONG committed:
Feat:
- 삭제한것도 커밋해야하는줄 몰랐습니다.

SEUNGBOONG committed:
Feat:
-  상위클래스에서 run하라는것이 어떤의미일까요?
  public static void main(String[] args) 이것을 쓰지 말라는 것일까요?

SEUNGBOONG committed:
Feat:
-  Null이 나오지않게 수정했습니다.
Collections.singletonList를 인텔리제이에선 지향한다고 하길래 찾아서 적용해봤습니다.
[URL](https://www.korecmblog.com/blog/arrays-asList-alternative-Collections-singletonList) 근데 아직 이해가 되진않습니다. 이것이 빅오n관련 내용일까요>

SEUNGBOONG committed:
Feat:

SEUNGBOONG committed:
Feat:
-불필요한 get 삭제

SEUNGBOONG committed:
Feat:
- 대문자 소문자 수정

SEUNGBOONG committed:
Fix:
- 알려주신 assertAll적용

sosow0212 commented on src/test/java/domain/LottoTest.java:
```diff
@@ -23,9 +24,11 @@ class LottoTest {
-        assertEquals(expectedValue, generatedNumbers);
-        assertEquals(6, generatedNumbers.size());
-        assertEquals(List.of(5, 16, 11, 23, 32, 44), generatedNumbers);
+        Assertions.assertAll(
```
👍 softly와 차이점도 알면 좋을 것 같아요

SEUNGBOONG replied:
softly공부해서 멘토님 앞에서 설명해보겠씁니다.

sosow0212 reviewed (approved):
수고하셨습니다!! 조금 부족한 부분은 다음 미션하면서 채워나가도 될 것 같아요.
말씀하신 부분들은 오프라인으로 설명하도록 하겠습니다!!!
이번 미션도 수고하셨습니다~~ 👏
